### ISSUE TITLE
About the training of SD2.1-VAE16

### ISSUE THREAD

**Issue by CuddleSabe** [icandle/GenDR#2]
Thanks for your Great Work !
I'm recurrenting the training of **SD2.1-VAE16**.
I use 8 A100 for bz=2048， the results of 3w iters is bad, and the diff loss is about 0.12, is it normal?

<img width="932" alt="Image" src="https://github.com/user-attachments/assets/fd951d05-1051-4cf7-b343-6c739e2aac83" />

<img width="317" alt="Image" src="https://github.com/user-attachments/assets/96a16df0-ec73-4e59-b82e-e3fd8f7fd0e7" />

Can you reshare the training log ? Thanks !

**Comment by CuddleSabe** [comment]
<img width="312" alt="Image" src="https://github.com/user-attachments/assets/d33d8e76-dbc4-40e3-9d0f-91ed93aa6103" />

and this is the log of REPA loss, I use the dino output (bx768x32x32) to supervise the first encoder of the unet (bx320x32x32), the mlp is construct with k=1 conv2d and silu ( c320->c1024->c768)

**Comment by CuddleSabe** [comment]
<img width="553" alt="Image" src="https://github.com/user-attachments/assets/ef26db4d-4735-4093-b584-897b28857caa" />

**Comment by icandle** [comment]
Hi, the diffusion loss seems normal, it does converge around 0.1. Will changing sampling scheduler or adding negtive prompt help generate a better results? 

For REPA, our cosine similarity is converge around 0.7. The REPA loss might be larger than ours, leading to unet overfitting on REPA loss than diffusion loss, I'm not sure, maybey you can removing REPA or adjusting coefficient for validation. Besides, the REPA module in our paper is realized as:

```
        projector_dim = 2048
        z_dim = 768
        self.projectors = nn.Sequential(
                PatchEmbed(32, 1, 320, 1152, bias=True ),
                nn.Linear(1152, projector_dim),
                nn.SiLU(),
                nn.Linear(projector_dim, projector_dim),
                nn.SiLU(),
                nn.Linear(projector_dim, z_dim),
            )
```

Moreover, the losses are realized follow REPA:
```
            z = torch.nn.functional.normalize(z, dim=-1)
            z_diffusion = torch.nn.functional.normalize(z_diffusion, dim=-1)

            proj_loss = mean_flat(-(z * z_diffusion).sum(dim=-1)).mean()

            loss = denoising_loss + 0.01*proj_loss
```

**Comment by icandle** [comment]
> <img alt="Image" width="553" src="https://private-user-images.githubusercontent.com/61224076/435520013-ef26db4d-4735-4093-b584-897b28857caa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyMDY3NzMsIm5iZiI6MTc0NTIwNjQ3MywicGF0aCI6Ii82MTIyNDA3Ni80MzU1MjAwMTMtZWYyNmRiNGQtNDczNS00MDkzLWI1ODQtODk3YjI4ODU3Y2FhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIxVDAzMzQzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIwZDlmYzZkZDEwZmE0M2FiODA3MWQ0ZjhiOTY0Zjc1MDEyYTc0NTc5ZDg2NGU4ZTdjMzMyMzZkZmMwZDk4MzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZZOHZMwxUPQeuOfQxF8ycRkOvDmu7DZR_fF6tSskDIg">

We directly change the latent channel to 16 in config and allow paramter load  mismatching. This implementation is even better.

**Comment by CuddleSabe** [comment]
> > <img alt="Image" width="553" src="https://private-user-images.githubusercontent.com/61224076/435520013-ef26db4d-4735-4093-b584-897b28857caa.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDUyMDY3NzMsIm5iZiI6MTc0NTIwNjQ3MywicGF0aCI6Ii82MTIyNDA3Ni80MzU1MjAwMTMtZWYyNmRiNGQtNDczNS00MDkzLWI1ODQtODk3YjI4ODU3Y2FhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MjElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDIxVDAzMzQzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIwZDlmYzZkZDEwZmE0M2FiODA3MWQ0ZjhiOTY0Zjc1MDEyYTc0NTc5ZDg2NGU4ZTdjMzMyMzZkZmMwZDk4MzkmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.ZZOHZMwxUPQeuOfQxF8ycRkOvDmu7DZR_fF6tSskDIg">
> 
> We directly change the latent channel to 16 in config and allow paramter load mismatching. This implementation is even better.

Thanks for your reply !!
I will try it :)

**Comment by CuddleSabe** [comment]
> Hi, the diffusion loss seems normal, it does converge around 0.1. Will changing sampling scheduler or adding negtive prompt help generate a better results?
> 
> For REPA, our cosine similarity is converge around 0.7. The REPA loss might be larger than ours, leading to unet overfitting on REPA loss than diffusion loss, I'm not sure, maybey you can removing REPA or adjusting coefficient for validation. Besides, the REPA module in our paper is realized as:
> 
> ```
>         projector_dim = 2048
>         z_dim = 768
>         self.projectors = nn.Sequential(
>                 PatchEmbed(32, 1, 320, 1152, bias=True ),
>                 nn.Linear(1152, projector_dim),
>                 nn.SiLU(),
>                 nn.Linear(projector_dim, projector_dim),
>                 nn.SiLU(),
>                 nn.Linear(projector_dim, z_dim),
>             )
> ```
> 
> Moreover, the losses are realized follow REPA:
> 
> ```
>             z = torch.nn.functional.normalize(z, dim=-1)
>             z_diffusion = torch.nn.functional.normalize(z_diffusion, dim=-1)
> 
>             proj_loss = mean_flat(-(z * z_diffusion).sum(dim=-1)).mean()
> 
>             loss = denoising_loss + 0.01*proj_loss
> ```

Hi, which lr_scheduler you used ? constant 1e-5 or cosine, thanks !

**Comment by CuddleSabe** [comment]
> Hi, the diffusion loss seems normal, it does converge around 0.1. Will changing sampling scheduler or adding negtive prompt help generate a better results?
> 
> For REPA, our cosine similarity is converge around 0.7. The REPA loss might be larger than ours, leading to unet overfitting on REPA loss than diffusion loss, I'm not sure, maybey you can removing REPA or adjusting coefficient for validation. Besides, the REPA module in our paper is realized as:
> 
> ```
>         projector_dim = 2048
>         z_dim = 768
>         self.projectors = nn.Sequential(
>                 PatchEmbed(32, 1, 320, 1152, bias=True ),
>                 nn.Linear(1152, projector_dim),
>                 nn.SiLU(),
>                 nn.Linear(projector_dim, projector_dim),
>                 nn.SiLU(),
>                 nn.Linear(projector_dim, z_dim),
>             )
> ```
> 
> Moreover, the losses are realized follow REPA:
> 
> ```
>             z = torch.nn.functional.normalize(z, dim=-1)
>             z_diffusion = torch.nn.functional.normalize(z_diffusion, dim=-1)
> 
>             proj_loss = mean_flat(-(z * z_diffusion).sum(dim=-1)).mean()
> 
>             loss = denoising_loss + 0.01*proj_loss
> ```

and which sample scheduler in your training ? Euler-base or ddpm? thanks!

**Comment by icandle** [comment]
The lr is constant. We train with DDPM and test with Euler. Specifically, we use "epsilon" prediction rather than "v_prediction" in original SD2.1.

**Comment by CuddleSabe** [comment]
> The lr is constant. We train with DDPM and test with Euler. Specifically, we use "epsilon" prediction rather than "v_prediction" in original SD2.1.

Thanks..the results of 5w iter are still very noisy... could your share your validation samples in the training? ;)

![Image](https://github.com/user-attachments/assets/e01759a3-3b3d-4175-b834-80586df197b1)

![Image](https://github.com/user-attachments/assets/1e459f4d-2961-49a9-9074-7b8c81a794a6)

![Image](https://github.com/user-attachments/assets/080d02f7-2fa1-4bc9-ba9f-074b0a2b0bb1)

**Comment by CuddleSabe** [comment]
> The lr is constant. We train with DDPM and test with Euler. Specifically, we use "epsilon" prediction rather than "v_prediction" in original SD2.1.

Hi, which "mixed_precision" do you use?

**Comment by icandle** [comment]
> > The lr is constant. We train with DDPM and test with Euler. Specifically, we use "epsilon" prediction rather than "v_prediction" in original SD2.1.
> 
> Hi, which "mixed_precision" do you use?

We utilize bf16. The validation samples in the training (10k,20k,30k) for prompt "a golden vase with different flowers" are:

![Image](https://github.com/user-attachments/assets/74ec0b3e-2c32-41a3-a927-a7a6bbcd1ad2)

![Image](https://github.com/user-attachments/assets/843d40ce-c37e-4948-86d9-efd2f42f1e18)

![Image](https://github.com/user-attachments/assets/c0280327-ad0d-4cff-aec6-a38e96e68c24)

**Comment by CuddleSabe** [comment]
> > > The lr is constant. We train with DDPM and test with Euler. Specifically, we use "epsilon" prediction rather than "v_prediction" in original SD2.1.
> > 
> > 
> > Hi, which "mixed_precision" do you use?
> 
> We utilize bf16. The validation samples in the training (10k,20k,30k) for prompt "a golden vase with different flowers" are:
> 
> ![Image](https://github.com/user-attachments/assets/74ec0b3e-2c32-41a3-a927-a7a6bbcd1ad2)
> 
> ![Image](https://github.com/user-attachments/assets/843d40ce-c37e-4948-86d9-efd2f42f1e18)
> 
> ![Image](https://github.com/user-attachments/assets/c0280327-ad0d-4cff-aec6-a38e96e68c24)

sad....we use the flux vae,  I don't know the reason...

**Comment by icandle** [comment]
> sad....we use the flux vae, I don't know the reason...

Flux VAE adds shift_factor, this might be the reason for the noise. We use [Ostris-VAE](https://huggingface.co/ostris/vae-kl-f8-d16), which has the same architecture as VAE of SD21 but change the latent channels.

**Comment by CuddleSabe** [comment]
> > sad....we use the flux vae, I don't know the reason...
> 
> Flux VAE adds shift_factor, this might be the reason for the noise. We use [Ostris-VAE](https://huggingface.co/ostris/vae-kl-f8-d16), which has the same architecture as VAE of SD21 but change the latent channels.

thanks! i will try

**Comment by CuddleSabe** [comment]
> > sad....we use the flux vae, I don't know the reason...
> 
> Flux VAE adds shift_factor, this might be the reason for the noise. We use [Ostris-VAE](https://huggingface.co/ostris/vae-kl-f8-d16), which has the same architecture as VAE of SD21 but change the latent channels.

Hi, the "encoded output h_t" is the output of the first down_block in unet, or the input/output of the mid_block? thanks

<img width="577" alt="Image" src="https://github.com/user-attachments/assets/3df8a4b6-c865-4df6-9234-dd49e7d276b4" />

**Comment by icandle** [comment]
> Hi, the "encoded output h_t" is the output of the first down_block in unet, or the input/output of the mid_block? thanks


"encoded output h_t" is the output of the first down_block in unet.

**Comment by CuddleSabe** [comment]
> > Hi, the "encoded output h_t" is the output of the first down_block in unet, or the input/output of the mid_block? thanks
> 
> "encoded output h_t" is the output of the first down_block in unet.

Hi, i trained on lora(r=64) with lr=5e-5, and it has normal results! thanks

<img width="819" alt="Image" src="https://github.com/user-attachments/assets/ad28c13f-169b-4001-bbb3-2292a9bc8506" />

but it worse than yours...lol
